src/core/lib/event_engine/ares_driver.cc

```diff
@@ -0,0 +1,1177 @@
+namespace {
+
+struct FdNode;
+class FdNodeList;
+
+class GrpcAresRequestImpl : public grpc_core::RefCounted<GrpcAresRequestImpl> {
+ public:
+  GrpcAresRequestImpl(absl::string_view name,
+                      absl::optional<absl::string_view> default_port,
+                      EventEngine::Duration timeout,
+                      RegisterAresSocketWithPollerCallback register_cb,
+                      EventEngine* event_engine);
+  ~GrpcAresRequestImpl() override;
+
+  absl::Status Initialize(absl::string_view dns_server, bool check_port)
+      ABSL_LOCKS_EXCLUDED(mu_);
+  bool Cancel() ABSL_LOCKS_EXCLUDED(mu_);
+  absl::string_view host() const ABSL_EXCLUSIVE_LOCKS_REQUIRED(mu_) {
+    return host_;
+  }
+  int port() const ABSL_EXCLUSIVE_LOCKS_REQUIRED(mu_) { return port_; }
+
+ protected:
+  void Work() ABSL_EXCLUSIVE_LOCKS_REQUIRED(mu_);
+  void StartTimers() ABSL_EXCLUSIVE_LOCKS_REQUIRED(mu_);
+  void CancelTimers() ABSL_EXCLUSIVE_LOCKS_REQUIRED(mu_);
+
+ private:
+  friend class GrpcPolledFd;
+  friend class GrpcPolledFdFactory;
+  friend struct FdNode;
+  friend class FdNodeList;
+
+  absl::Status SetRequestDNSServer(absl::string_view dns_server)
+      ABSL_EXCLUSIVE_LOCKS_REQUIRED(mu_);
+  void OnReadable(FdNode* fd_node, absl::Status status)
+      ABSL_LOCKS_EXCLUDED(mu_);
+  void OnWritable(FdNode* fd_node, absl::Status status)
+      ABSL_LOCKS_EXCLUDED(mu_);
+  void OnHandleDestroyed(FdNode* fd_node, absl::Status status)
+      ABSL_LOCKS_EXCLUDED(mu_);
+  void OnQueryTimeout() ABSL_LOCKS_EXCLUDED(mu_);
+  void OnAresBackupPollAlarm() ABSL_LOCKS_EXCLUDED(mu_);
+  void ShutdownPollerHandlesLocked(absl::Status status)
+      ABSL_EXCLUSIVE_LOCKS_REQUIRED(mu_);
+
+ protected:
+  grpc_core::Mutex mu_;
+  bool initialized_ ABSL_GUARDED_BY(mu_) = false;
+  /// name to resolve
+  const std::string name_ ABSL_GUARDED_BY(mu_);
+  const std::string default_port_ ABSL_GUARDED_BY(mu_);
+  // ares channel
+  ares_channel channel_ ABSL_GUARDED_BY(mu_) = nullptr;
+  /// host to resolve; parsed from the name to resolve
+  absl::string_view host_ ABSL_GUARDED_BY(mu_);
+  /// port; parsed from the name to resolve or default_port_
+  int port_ ABSL_GUARDED_BY(mu_) = 0;
+  const EventEngine::Duration timeout_ ABSL_GUARDED_BY(mu_);
+  size_t pending_queries_ ABSL_GUARDED_BY(mu_) = 0;
+  bool shutting_down_ ABSL_GUARDED_BY(mu_) = false;
+  bool cancelled_ ABSL_GUARDED_BY(mu_) = false;
+  absl::Status error_ ABSL_GUARDED_BY(mu_);
+  std::unique_ptr<FdNodeList> fd_node_list_ ABSL_GUARDED_BY(mu_);
+  EventEngine* event_engine_;
+  absl::optional<EventEngine::TaskHandle> query_timeout_handle_
+      ABSL_GUARDED_BY(mu_);
+  absl::optional<EventEngine::TaskHandle> ares_backup_poll_alarm_handle_
+      ABSL_GUARDED_BY(mu_);
+  std::unique_ptr<GrpcPolledFdFactory> polled_fd_factory_ ABSL_GUARDED_BY(mu_);
+};
+
+class GrpcAresHostnameRequestImpl final : public GrpcAresHostnameRequest,
+                                          public GrpcAresRequestImpl {
```

Comment:
Both `GrpcAresHostnameRequest` and `GrpcAresRequestImpl` have a `Cancel` method, but I think it's only implemented in `GrpcAresRequestImpl` today (though `GrpcAresRequestImpl` does not implement that interface). Have you considered instead making `GrpcAresHostnameRequestImpl` _have_ a base request, rather than _be_ a base request?

Again, I think multiple inheritance of _interfaces_ is usually fine, but any behavior inheritance gets messy. I'd like to see a design without it.